gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -587,19 +660,31 @@ class NoiseModelFactorN
    * e.g. `const Vector error = factor.evaluateError(pose, point);`
    */
   inline Vector evaluateError(const ValueTypes&... x) const {
-    return evaluateError(x..., OptionalMatrix<ValueTypes>()...);
+    return evaluateError(x..., OptionalMatrixTypeT<ValueTypes>()...);
   }
 
   /** Some (but not all) optional Jacobians are omitted (function overload)
-   *
+   * and the jacobians are l-value references to matrices.
    * e.g. `const Vector error = factor.evaluateError(pose, point, Hpose);`
    */
-  template <typename... OptionalJacArgs,
-            typename = IndexIsValid<sizeof...(OptionalJacArgs) + 1>>
-  inline Vector evaluateError(const ValueTypes&... x,
-                              OptionalJacArgs&&... H) const {
-    return evaluateError(x..., std::forward<OptionalJacArgs>(H)...,
-                         boost::none);
+  template <typename... OptionalJacArgs, typename = IndexIsValid<sizeof...(OptionalJacArgs) + 1>>
+  inline AreAllMatrixRefs<Vector, OptionalJacArgs...> evaluateError(const ValueTypes&... x,
+                                                                  OptionalJacArgs&&... H) const {
+    return evaluateError(x..., (&H)...);
+  }
+
+  /** Some (but not all) optional Jacobians are omitted (function overload)
+   * and the jacobians are pointers to matrices.
+   * e.g. `const Vector error = factor.evaluateError(pose, point, &Hpose);`
+   */
+  template <typename... OptionalJacArgs, typename = IndexIsValid<sizeof...(OptionalJacArgs) + 1>>
+  inline AreAllMatrixPtrs<Vector, OptionalJacArgs...> evaluateError(const ValueTypes&... x,
+                                                                    OptionalJacArgs&&... H) const {
+    // If they are pointer version, ensure to cast them all to be Matrix* types
+    // This will ensure any arguments inferred as std::nonetype_t are cast to (Matrix*) nullptr
+    // This guides the compiler to the correct overload which is the one that takes pointers
+    return evaluateError(x..., 
+        std::forward<OptionalJacArgs>(H)..., static_cast<OptionalMatrixType>(OptionalNone));
```

Comment:
May or may not be worth doing, but I think the pointer versions of these can be replaced with default args.  i.e.

```c++
template <typename... OptionalJacArgs, typename = IndexIsValid<sizeof...(OptionalJacArgs) + 1>>
virtual Vector evaluateError(const ValueTypes&... x, OptionalMatrixTypeT H1 = 0, OptionalJacArgs*... Hs) = 0;
```
or something along those lines (untested).  Or maybe I'm wrong, because otherwise I'm not sure why we did these weird overloading tricks previously.  Maybe I didn't realize that if you provide one default arg, the remaining template params automatically get default args.

Reply:
Any remaining optional arguments are added by the line 686, recursively, right?
```
return evaluateError(x..., 
        std::forward<OptionalJacArgs>(H)..., static_cast<OptionalMatrixType>(OptionalNone));
```

Reply:
Ya sorry I wasn't very clear, I meant to say that I think it might be more readable with the default arguments rather than the lots-of-templates version.  But nbd :)